### DISCUSSION TITLE
Is there a recommended way of enabling introspection on a per request basis?

### DISCUSSION THREAD

**Discussion by lolopinto** [mirumee/ariadne#1168] [General]
I want to enable introspection dynamically. what's the best way to do that?

right now, it seems introspection has to be enabled all or nothing and I'd like to check based on the request and then configure `GraphQL` based on that.

using my own handler and calling `configure` in each doesn't quite work because there could be multiple requests at the same time. The best way I have so far is to expose a different endpoint and a different instance of `GraphQL` there and perform the check in a custom `GraphQLHTTPHandler`'s `handle_request` method but that doesn't seem elegant.

**Comment by rafalp** [comment]
What do you mean by introspection by query? Do you want to let people introspect some of introspection fields but not others?

**Comment by lolopinto** [comment]
Sorry, should have said introspection by request.

I want to define introspection based on request. E.g. doing a launchdarkly
or statsig check based on IP

**Reply by rafalp** [reply]
I am afraid I still don't understand what you are after? "launchdarkly" and "statsig" doesn't tell me anything.

**Reply by lolopinto** [reply]
Right now, introspection is configured on a per-route basis and the introspection flag is passed to the http handler and all the way down to the graphql function.

So that value persists across all requests as far as I can tell.

I want to configure it on a per request basis. I'm asking how to do that.

**Reply by rafalp** [reply]
So you want third party services to be able to introspect your schema, but not the other clients?

**Reply by rafalp** [reply]
Leave introspection enabled, but use custom [`get_validation_rules`](https://ariadnegraphql.org/docs/query-validators#configuring-cost_validator-dynamically) to conditionally return `(IntrospectionDisabledRule, )` if `context["request"]` fails to match.

You can import `IntrospectionDisabledRule` from `ariadne.validation.introspection_disabled`

**Reply by lolopinto** [reply]
that works! thanks!
